core/src/main/java/com/graphhopper/routing/util/FlagEncoder.java

```diff
@@ -53,6 +54,14 @@
      */
     DecimalEncodedValue getAverageSpeedEnc();
 
+    /**
+     * Decide whether a restriction relation is enabled for a given mode of travel. This skips some relations before
+     * handleRelationTags is called.
+     *
+     * @return the encoded value to indicate if this encoder allows travel or not.
+     */
+    EncodingManager.RelationAcceptation getRelationAccept(ReaderRelation relation);
```

Comment:
Why return `RelationAcceptation` here ? Why not just a boolean ? All this method does is deciding whether or not the turn rrestriction should be applied for this encoder, or am I missing something ?
I am also wondering why this is about 'relations' in general since we are only using it to filter turn restrictions here. Also, I'm not a native English speaker, but I do not think `accept` is a noun ? What also bothers me a bit is that the parsing of the 'dedicated'/conditional turn restrictions is kind of spread across `OSMReader#createTurnRelation` and `CarFlagEncoder#hasRestrictedRelationTag`. What if we wanted to do the same for another encoder than car ? We would have to do the same parsing again. Can we not do this in `OSMReader` already and provide the parsed information (like the `except` list to the encoder) (probably by putting it into `OSMTurnRelation`) ?

For these reasons I think the signature of this method should rather be:

```java
public abstract boolean acceptsRelation(OSMTurnRelation relation);
```

Reply:
Ok, I think I understand how to correct it. I will try. I did that way to make similar with `Way`, but you're right, only a boolean is needed for `Relation`.

Reply:
I don't know how to deal with the class `OSMTurnRelation` (subproject reader-osm) used into `FlagEncoder` (subproject core), the dependency is in the other direction. Do I need to create a new class (for example `ReaderTurnRelation`) which extends `ReaderRelation` to store the parsed data (like the `except` list)? Or do I need to use the `members` list of `ReaderRelation` which doesn't seem appropriate? Thanks

Reply:
@karussell is there any reason why we should not move `OSMTurnRelation` into core (next to `ReaderRelation` etc.) ?

Reply:
I think I would just move it into core.

Reply:
`OSMTurnRelation` uses `OSMReader` in function `getRestrictionAsEntries`.

Reply:
Right, but that does not make much sense to me. `OSMTurnRelation` is an (intermediate) data class here. Extracting the `TurnCostTableEntry`s is part of the (OSM-parsing) algorithm which happens in `OSMReader`, so the dependency should be turned inside out: `OSMTurnRelation` should give access to its data and the `getRestrictionAsEntries` method should go into `OSMReader`. To make sure its still testable (as in `OSMTurnRelationTest`) maybe either we can make the method static or put this part of `OSMReader` into a separate class. Does that make sense ?

Reply:
This definitely needs some refactoring. The best change, likely in a separate issue, would be if it is decoupled from the FlagEncoder so that the CarFlagEncoder can use car turn restrictions but it can be also used from a BusFlagEncoder without the requirement to extend this class from the CarFlagEncoder (i.e. composition over inheritance).

And the other problem is that the parsing part is something OSM-specific that should not be in core. We have the same difference for e.g. the `max_weight`-EncodedValue vs. OSMMaxWeightParser. But even there the OSM-specific class is still in core.

Reply:
@sguill do you want to do this refactoring in a separate PR ? if not I can do it as well.

Reply:
I think this is good now. `OSMReader` does the osm parsing part and each encoder can decide what turn restrictions should apply for it (given the information about what vehicles where specified in the restriction tags (including the except list)). The flag encoders still contain a lot of stuff that is specific to OSM tags, but at least the parsing happens in OSMReader. What is still a bit ugly is that `OSMTurnRelation` contains osm way ids, while for the purpose here this is not needed. We could separate this further to make this class less osm specific, but not sure if this needs to be done here.